lib/std/math/nextafter.zig

```diff
@@ -0,0 +1,304 @@
+const builtin = @import("builtin");
+const std = @import("../std.zig");
+const math = std.math;
+const assert = std.debug.assert;
+const expect = std.testing.expect;
+
+/// Returns the next representable value after `x` in the direction of `y`.
+///
+/// Special cases:
+///
+/// - If `x == y`, `y` is returned.
+/// - For floating-point values, if either `x` or `y` is a NaN, a NaN is returned.
+/// - For floating-point values, if `x == 0.0` and `@abs(y) > 0.0`, the smallest subnormal number
+///   with the sign of `y` is returned.
+///
+pub fn nextAfter(comptime T: type, x: T, y: T) T {
+    return switch (@typeInfo(T)) {
+        .Int, .ComptimeInt => nextAfterInt(T, x, y),
+        .Float => nextAfterFloat(T, x, y),
+        else => @compileError("Expected int or non-comptime float, found '" ++ @typeName(T) ++ "'"),
+    };
+}
+
+fn nextAfterInt(comptime T: type, x: T, y: T) T {
+    comptime assert(@typeInfo(T) == .Int or @typeInfo(T) == .ComptimeInt);
+    return if (@typeInfo(T) == .Int and @typeInfo(T).Int.bits < 2)
+        // Special case for `i0`, `u0`, `i1`, and `u1`.
+        y
+    else if (y > x)
+        x + 1
+    else if (y < x)
+        x - 1
+    else
+        y;
+}
+
+// Based on nextafterf/nextafterl from mingw-w64 which are both public domain.
+// <https://github.com/mingw-w64/mingw-w64/blob/e89de847dd3e05bb8e46344378ce3e124f4e7d1c/mingw-w64-crt/math/nextafterf.c>
+// <https://github.com/mingw-w64/mingw-w64/blob/e89de847dd3e05bb8e46344378ce3e124f4e7d1c/mingw-w64-crt/math/nextafterl.c>
+
+fn nextAfterFloat(comptime T: type, x: T, y: T) T {
+    comptime assert(@typeInfo(T) == .Float);
+    const Bits = std.meta.Int(.unsigned, @typeInfo(T).Float.bits);
+    if (x == y) {
+        // Returning `y` ensures that (0.0, -0.0) returns -0.0 and that (-0.0, 0.0) returns 0.0.
+        return y;
+    }
+    if (math.isNan(x) or math.isNan(y)) {
+        return math.nan(T);
+    }
+    if (x == 0.0) {
+        const sign_bit_mask = @as(Bits, 1) << (@bitSizeOf(T) - 1);
+        const smallest_bits: Bits = @bitCast(math.floatTrueMin(T));
+        const y_sign_bit = @as(Bits, @bitCast(y)) & sign_bit_mask;
+        return @bitCast(smallest_bits | y_sign_bit);
+    }
```

Comment:
I can think of at least three ways of returning +-smallest here but I don't know which way is the best and easiest for the compiler to optimize.

```zig
const std = @import("std");
const math = std.math;

export fn copySign(y: f32) f32 {
    const sign_bit_mask: u32 = 1 << (@bitSizeOf(f32) - 1);
    const smallest_bits: u32 = @bitCast(math.floatTrueMin(f32));
    const y_sign_bit = @as(u32, @bitCast(y)) & sign_bit_mask;
    return @bitCast(smallest_bits | y_sign_bit);
}

export fn compareFloat(y: f32) f32 {
    return if (y > 0.0)
        math.floatTrueMin(f32)
    else
        -math.floatTrueMin(f32);
}

export fn compareSignedInt(y: f32) f32 {
    const y_bits_signed: i32 = @bitCast(y);
    return if (y_bits_signed >= 0)
        math.floatTrueMin(f32)
    else
        -math.floatTrueMin(f32);
}
```

```asm
copySign:
        vmovmskps       eax, xmm0
        shl     eax, 31
        or      eax, 1
        vmovd   xmm0, eax
        ret

.LCPI1_0:
        .long   0x80000001
.LCPI1_1:
        .long   0x00000001
compareFloat:
        vxorps  xmm1, xmm1, xmm1
        vcmpltss        k1, xmm1, xmm0
        vmovss  xmm0, dword ptr [rip + .LCPI1_0]
        vmovss  xmm0 {k1}, dword ptr [rip + .LCPI1_1]
        ret

.LCPI2_0:
        .long   0x00000001
.LCPI2_1:
        .long   0x80000000
compareSignedInt:
        vpbroadcastd    xmm1, dword ptr [rip + .LCPI2_0]
        vpternlogd      xmm0, xmm1, dword ptr [rip + .LCPI2_1]{1to4}, 236
        ret
```

https://godbolt.org/z/x6zevjd4T

Reply:
I thought about it some more and decided to change this into the `if (y > 0.0) trueMin else -trueMin` variant. All three options are branchless (which is probably the most important thing) and out of the three it's the most immediately obvious and readable one.